File: app/src/main/java/com/game/tama/command/CommandQueue.java
package com.game.tama.command;

import android.util.Log;

import com.game.engine.DisplayAdapter;
import com.game.tama.core.World;
import com.game.tama.core.WorldObject;
import com.game.tama.thing.pet.Pet;
import com.game.tama.util.Vec2;

import java.util.LinkedList;
import java.util.Queue;

/**
 * A self managing list commands that a pet will execute one at a time,
 * regardless of command failures unless failAllOnFail is set to true
 */
public class CommandQueue extends Command
{
    private final Queue<Command> queue;
    private Runnable failAction = () ->
    {
    };
    private PathInit initializer = null;
    public WorldObject ultimateTarget = null;

    public CommandQueue()
    {
        this((final CommandQueue queue, final World world, final Pet pet) ->
             {
                 Log.d("Command queue", "default initiliser");
             });
    }

    public CommandQueue(final PathInit initializer)
    {
        super();
        this.initializer = initializer;
        update = this::start;
        queue = new LinkedList<>();
        failAction = () ->
        {
            queue.poll();
        };
    }

    @Override
    public void start(final Pet pet, final World world)
    {
        Log.d(this.getClass().getCanonicalName(), "starting queue");
        super.start(pet, world);
        initializer.run(this, world, pet);
    }

    @Override
    public void doing(final Pet pet, final World world)
    {
        final Command current = queue.peek();
        if (current == null)
        {
            state = CommandState.complete;
            return;
        }
        current.update.update(pet, world);
        switch (current.state)
        {
            case complete:
                queue.poll();
                break;
            case failed:
            {
                Log.d(
                    this.getClass().getCanonicalName(),
                    "failed queue element");
                failAction.run();
                break;
            }
        }
    }

    public void failAllOnFail(final boolean failAll)
    {
        if (failAll)
        {
            failAction = () ->
            {
                state = CommandState.failed;
            };
        }
        else
        {
            failAction = () ->
            {
                queue.poll();
            };
        }
    }

    public int length()
    {
        return queue.size();
    }

    public void add(final Command command)
    {
        queue.add(command);
    }

    @Override
    public void draw(final DisplayAdapter d)
    {
        if (actor == null || ultimateTarget == null)
        {
            return;
        }
        final Vec2<Float> start = actor.loc.getWorldArrPos();
        start.x += 1;
        start.y += 1;
        final Vec2<Float> end = ultimateTarget.getWorldArrPos();
        end.x += 1;
        end.y += 1;
        d.drawLine(
            (start.x - 0.5f),
            (start.y - 0.5f),
            (end.x - 0.5f),
            (end.y - 0.5f));
    }

    @Override
    public boolean isReplaceable()
    {
        final Command current = queue.peek();
        return current == null || current.isReplaceable();
    }

    @Override
    public void hardCancel()
    {
        final Command command = queue.peek();
        command.hardCancel();
        super.hardCancel();
    }
}

File: app/src/main/java/com/game/tama/command/CommandQueue.kt
package com.game.tama.command

import android.util.Log
import com.game.android.DisplayAdapter
import com.game.tama.thing.Pet
import com.game.tama.core.World
import com.game.tama.core.WorldObject
import com.game.tama.util.Vec2
import java.util.*

/**
 * A self managing list commands that a pet will execute one at a time,
 * regardless of command failures unless failAllOnFail is set to true
 */
class CommandQueue constructor() : Command()
{
    private val queue: Queue<Command>;
    private var failAction: () -> Unit;
    private var initialiser: (CommandQueue, World, Pet) -> Unit =
            { queue, world, pet ->
                Log.d(this.javaClass.canonicalName, "default initiliser");
            };
    var ultimateTarget: WorldObject? = null;

    constructor(initializer: (CommandQueue, World, Pet) -> Unit) : this()
    {
        Log.d(this.javaClass.canonicalName, "initilizing queue");
        this.initialiser = initializer;
    }

    init
    {
        update = ::start;
        queue = LinkedList()
        failAction = { queue.poll() }
    }

    override fun start(pet: Pet, world: World)
    {
        super.start(pet, world);
        Log.d(this.javaClass.canonicalName, "starting queue");
        initialiser(this, world, pet);
    }

    override fun doing(pet: Pet, world: World): Unit
    {
        val current = queue.peek();
        if (current == null)
        {
            state = CommandState.complete;
            return;
        }
        current.update(pet, world)
        when (current.state)
        {
            CommandState.complete -> queue.poll();
            CommandState.failed   ->
            {
                Log.d(this.javaClass.canonicalName, "failed queue element");
                failAction()
            };
        }
    }

    fun failAllOnFail(failAll: Boolean)
    {
        when (failAll)
        {
            true  -> failAction = { state = CommandState.failed }
            false -> failAction = { queue.poll() }
        }
    }

    fun length(): Int
    {
        return queue.size
    }

    fun add(command: Command)
    {
        queue.add(command);
    }

    override fun draw(d: DisplayAdapter)
    {
        if (actor == null || ultimateTarget == null)
            return;
        val start: Vec2<Float> = actor!!.loc.worldArrPos;
        start.x += 1;
        start.y += 1;
        val end: Vec2<Float> = ultimateTarget!!.worldArrPos;
        end.x += 1;
        end.y += 1;
        d.drawLine((start.x * 16 - 8),
                   (start.y * 16 - 8),
                   (end.x * 16 - 8),
                   (end.y * 16 - 8));
    }

    override fun isReplaceable(): Boolean
    {
        val current: Command? = queue.peek();
        return current == null || current.isReplaceable();
    }

    override fun hardCancel()
    {
        val command: Command? = queue.peek();
        command?.hardCancel();
        super.hardCancel();
    }
}